File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(new ArrayList<>(), nums, ans, 0);
        return ans;

    }

    public void backtrack(List<Integer> cur, int[] nums, List<List<Integer>> ans, int index) {
        if (index > nums.length) {
            return;
        }
        ans.add(new ArrayList<>(cur));

        for (int i = index; i < nums.length; i++) {
            cur.add(nums[i]);
            // System.out.println("a" + index);
            backtrack(cur, nums, ans, i + 1);
            cur.remove(cur.size() - 1);
        }
    }
}

File: 78-subsets/subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    let ans = [];
    const n = nums.length;

    let backtrack = (cur, index) => {
        ans.push([...cur]);
        if(index >= n){
            return;
        }
        for(let i = index; i < n; i++){
            cur.push(nums[i]);
            backtrack(cur, i + 1);
            cur.pop();
        }
    }
    backtrack([], 0);
    return ans;
};